streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredSessionStore.java

```diff
@@ -0,0 +1,169 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.StreamsMetrics;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.StateStore;
+import org.apache.kafka.streams.processor.internals.ProcessorStateManager;
+import org.apache.kafka.streams.state.KeyValueIterator;
+import org.apache.kafka.streams.state.SessionStore;
+import org.apache.kafka.streams.state.StateSerdes;
+
+import java.util.Objects;
+
+public class MeteredSessionStore<K, V> extends WrappedStateStore.AbstractStateStore implements SessionStore<K, V> {
+    private final SessionStore<Bytes, byte[]> inner;
+    private final String metricScope;
+    private final Serde<K> keySerde;
+    private final Serde<V> valueSerde;
+    private final Time time;
+    private StateSerdes<K, V> serdes;
+    private StreamsMetrics metrics;
+    private Sensor putTime;
+    private Sensor fetchTime;
+    private Sensor flushTime;
+    private Sensor removeTime;
+
+    MeteredSessionStore(final SessionStore<Bytes, byte[]> inner,
+                        final String metricScope,
+                        final Serde<K> keySerde,
+                        final Serde<V> valueSerde,
+                        final Time time) {
+        super(inner);
+        this.inner = inner;
+        this.metricScope = metricScope;
+        this.keySerde = keySerde;
+        this.valueSerde = valueSerde;
+        this.time = time;
+    }
+
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        //noinspection unchecked
+        this.serdes = new StateSerdes<>(ProcessorStateManager.storeChangelogTopic(context.applicationId(), name()),
+                                        keySerde == null ? (Serde<K>) context.keySerde() : keySerde,
+                                        valueSerde == null ? (Serde<V>) context.valueSerde() : valueSerde);
+        final String tagKey = "task-id";
+        final String tagValue = context.taskId().toString();
+        this.metrics = context.metrics();
+        this.putTime = this.metrics.addLatencyAndThroughputSensor(metricScope, name(), "put",
```

Comment:
As discussed on the KIP, this will also change the metrics a bit as we are not metering on each segment any more but on the higher level store wrapper.

Could we make the doc changes correspondingly in this PR as well so we do not need to come back later, as the main doc changes of KIP-182 is around public APIs so we may forget later.

Reply:
@guozhangwang we were never metering on each segment. The metric names haven't changed

Reply:
Hmm, you are right. The removing `MeteredSegmentedBytesStore` is always used as the wrapper for session stores.

BTW I just realized that it has one metric that would never be called, i.e. `get` operation. This should be OK as it should never be called before..